designs/generics-design.md

```diff
@@ -0,0 +1,403 @@
+```
+
+A type parameter of a generic class can also be instantiated with a user-defined
+class.
+
+5. **Type Variables and Generic Classes with Constraints**
+
+```python
+T = TypeVar('T')
+R = TypeVar('R', int, bool)
+
+class Box(Generic[T]):
+    pass
+
+class ConstrainedBox(Generic[R]):
+    pass
+
+b1: Box[int] = Box()
+cb1: ConstrainedBox[bool] = ConstrainedBox()
+```
+
+A type variable definition may include a list of types that constrains the types
+that the variable can be instantiated with. The list of types are interpreted as
+"OR" constraints on the type variable.
+
+6. **Generic Class with Fields of a Generic Type**
+```python
+T = TypeVar('T')
+
+class Box(Generic[T]):
+    f : T = __ZERO__
+
+b : Box[int] = None
+b = Box()
+print(b.f) #prints 0 (zero value of int)
+b.f = 10
+print(b.f) #prints 10
+```
+
+Generic classes can have fields with the type parameter as a type annotation. Since
+fields need to always be initialized but we don't know the actual type of the field, generic
+fields need to be initialized with the special __ZERO__ literal value that automatically uses
+a reasonable default value depending on the type the type-variable is actually instantiated with.
+
+7. **Generic Class with Methods with Generic Types in their signature**
+```python
+T = TypeVar('T')
+
+class Box(Generic[T]):
+    f : T = __ZERO__
+
+    def getF(self: Box[T]) -> T:
+        return self.f
+
+    def setF(self: Box[T], f: T):
+        self.f = f
+
+b : Box[int] = None
+b = Box()
+print(b.getF()) #prints 0 (zero value of int)
+b.setF(10)
+print(b.getF()) #prints 10
+print(b.f) #prints 10
+```
+
+Generic classes can have methods that use type parameters in their argument and
+return types. Also note the `Box[T]` annotation on the self argument.
+
+
+8. **Type Variables cannot shadow class name**
+```python
+T = TypeVar('T')
+
+class T(): # TypeError: Duplicate identifier in the same scope.
+    pass
+```
+
+Type Variable names cannot overlap with class, function or variable names. This
+is because in some contexts we need to be able to unambiguosly determine if an
+identifier is a type variable.
+
+9. **Generic Class with multiple Type Parameters**
+```python
+L = TypeVar('L')
+R = TypeVar('R')
+
+class Pair(Generic[L, R]):
+    left: L = __ZERO__
+    right: R = __ZERO__
+
+p1 : Pair[int, int] = None
+p1 = Pair()
+p2 : Pair[int, bool] = None
+p2 = Pair()
+```
+
+A class can have multiple type parameters. These type parameters can be independently
+instantiated to different types based on their own constraints.
+
+10. **Function that uses Type Parameters**
+```python
+T = TypeVar('T')
+
+def id(t: T) -> T:
+    return t
+```
+
+An independent function can also be generic. The arguments and return type can use
+any of the type variables that are declared in the program.
+
+11. **Joe's blessed test case from class**
+```python
+T= TypeVar('T')
+class C(Generic[T]):
+    t: T = __ZERO__
+    def f(self: C[T], other: C[T]):
+        other.t = self.t
```

Comment:
Right, and here the type being annotated correctly on `self`/`other` in the end will be important for typed loads and stores with memory management/offsets.